pkg/controllers/provisioner/v1alpha1/controller.go

```diff
@@ -61,17 +64,48 @@ func (c *Controller) Reconcile(object controllers.Object) error {
 
 	unschedulable := []*v1.Pod{}
 	for _, pod := range pods.Items {
+		tempPod := pod
 		for _, condition := range pod.Status.Conditions {
 			if condition.Type == v1.PodScheduled && condition.Reason == v1.PodReasonUnschedulable {
-				unschedulable = append(unschedulable, &pod)
+				unschedulable = append(unschedulable, &tempPod)
 			}
 		}
 	}
-
-	// 4. Attempt to schedule remaining pods by creating a set of nodes
-	if err := c.Allocator.Allocate(unschedulable); err != nil {
+	pendingPods := c.removeProcessedPods(unschedulable)
+	if len(pendingPods) == 0 {
+		return nil
+	}
+	// Attempt to schedule remaining pods by creating a set of nodes
+	if err := c.Allocator.Allocate(pendingPods); err != nil {
 		return fmt.Errorf("failed to allocate %d pods, %w", len(unschedulable), err)
 	}
-
 	return nil
 }
+
+func (c *Controller) removeProcessedPods(unschedulable []*v1.Pod) []*v1.Pod {
+	needsProcessing := make([]*v1.Pod, 0)
+	c.Lock()
+	defer c.Unlock()
+	for _, pod := range unschedulable {
+		key := podKeyCreate(pod)
+		// if a pod is never seen before, add to the list needsProcessing
+		if _, ok := c.ProcessedPods[key]; !ok {
+			needsProcessing = append(needsProcessing, pod)
+		}
+		// unschedulable pods are set as false
+		c.ProcessedPods[key] = false
+	}
+	// remove the pods which are not pending anymore.
+	for podKey, done := range c.ProcessedPods {
+		if done {
+			delete(c.ProcessedPods, podKey)
+			continue
+		}
+		c.ProcessedPods[podKey] = true
```

Comment:
What do you think about the following.
1. The allocator returns pods that it could and couldn't provision for
2. For the ones it did provision for, add them to "processed"
3. On each loop, filter out pods that are processed

Reply:
This needs more discussion, what scenarios do you think can occur when Allocator will not be able to provision? There could be API call failures and we need to do some work there to know what do in that case.